File: codex_poting/java/src/main/java/com/example/demo/mybatis/mapper/BookCustomMapper.java
package com.example.demo.mybatis.mapper;

import com.example.demo.mybatis.entity.BookWithPublisherName;
import com.example.demo.mybatis.generator.entity.BookSalesUnitPriceHistoryEntity;
import com.example.demo.mybatis.generator.entity.BookEntity;
import org.apache.ibatis.annotations.Mapper;
import org.apache.ibatis.annotations.Param;

import java.time.LocalDate;
import java.util.List;

@Mapper
public interface BookCustomMapper {
    BookWithPublisherName selectByPrimaryKeyWithPublisherName(@Param("id") Long id);

    List<BookWithPublisherName> selectByTitleOrAuthorStartingWithIgnoreCase(
        @Param("keyword") String keyword,
        @Param("releaseDateFrom") LocalDate releaseDateFrom,
        @Param("releaseDateTo") LocalDate releaseDateTo,
        @Param("limit") int limit,
        @Param("offset") long offset
    );

    long countByTitleOrAuthorStartingWithIgnoreCase(
        @Param("keyword") String keyword,
        @Param("releaseDateFrom") LocalDate releaseDateFrom,
        @Param("releaseDateTo") LocalDate releaseDateTo
    );

    BookEntity selectByPrimaryKeyWithWriteLock(@Param("id") Long id);

    void insertWithGeneratedKey(BookEntity row);

    Long selectNextSalesUnitPriceHistoryId();

    void insertSalesUnitPriceHistoryWithId(BookSalesUnitPriceHistoryEntity row);
}

File: codex_poting/kotlin/src/main/kotlin/com/example/demo/mybatis/mapper/BookCustomMapper.kt
package com.example.demo.mybatis.mapper

import com.example.demo.mybatis.entity.BookWithPublisherName
import com.example.demo.mybatis.generator.entity.BookEntity
import com.example.demo.mybatis.generator.entity.BookSalesUnitPriceHistoryEntity
import org.apache.ibatis.annotations.Mapper
import org.apache.ibatis.annotations.Param
import java.time.LocalDate

@Mapper
interface BookCustomMapper {
    fun selectByPrimaryKeyWithPublisherName(@Param("id") id: Long?): BookWithPublisherName?

    fun selectByTitleOrAuthorStartingWithIgnoreCase(
        @Param("keyword") keyword: String?,
        @Param("releaseDateFrom") releaseDateFrom: LocalDate?,
        @Param("releaseDateTo") releaseDateTo: LocalDate?,
        @Param("limit") limit: Int,
        @Param("offset") offset: Long
    ): MutableList<BookWithPublisherName?>?

    fun countByTitleOrAuthorStartingWithIgnoreCase(
        @Param("keyword") keyword: String?,
        @Param("releaseDateFrom") releaseDateFrom: LocalDate?,
        @Param("releaseDateTo") releaseDateTo: LocalDate?
    ): Long

    fun selectByPrimaryKeyWithWriteLock(@Param("id") id: Long?): BookEntity?

    fun insertWithGeneratedKey(row: BookEntity?)

    fun selectNextSalesUnitPriceHistoryId(): Long?

    fun insertSalesUnitPriceHistoryWithId(row: BookSalesUnitPriceHistoryEntity?)
}
